### PR TITLE
[Bugfix] Fix bug alter materialized view in concurrency

### LINKED ISSUE #10740: replay alter materialized view refresh cause NPE
### Steps to reproduce the behavior (Required)

2022-09-01 22:35:01,069 ERROR (replayer|72) [EditLog.loadJournal():943] Operation Type 10092
java.lang.NullPointerException: null
        at com.starrocks.alter.Alter.replayChangeMaterializedViewRefreshScheme(Alter.java:362) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayChangeMaterializedViewRefreshScheme(GlobalStateMgr.java:2715) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.EditLog.loadJournal(EditLog.java:269) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1702) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$5.runOneCycle(GlobalStateMgr.java:1568) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$5.run(GlobalStateMgr.java:1631) [starrocks-fe.jar:?]

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by Astralidea**
## What type of PR is this：
- [x] bugfix
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #10740

## Problem Summary(Required) ：

db.writeLock() have concurrency problem. we should use db.writeLockAndCheckExist() to avoid.
## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 7 / 12 (58.33%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|3|6|50.00%|[131, 164, 229]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|4|6|66.67%|[152, 548]|

**Review by ABingHuang — DISMISSED** [review]
LGTM

File: fe/fe-core/src/main/java/com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java [hunk 1]
```diff
@@ -148,7 +148,9 @@ public void processTaskRun(TaskRunContext context) throws Exception {
 
     private void updateMeta(ExecPlan execPlan) {
         // update the meta if succeed
-        database.writeLock();
+        if (!database.writeLockAndCheckExist()) {
+            throw new DmlException("update meta failed. database:" + database.getFullName() + " not exist");
+        }
```
Comment (gengjun-git) [inline fe/fe-core/src/main/java/com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java:153]: The table or partition may be dropped too. So you should check the existence of table/partition before update.

**Review by gengjun-git — DISMISSED** [review]

File: fe/fe-core/src/main/java/com/starrocks/alter/Alter.java [hunk 2]
```diff
@@ -127,7 +127,9 @@ public void processCreateMaterializedView(CreateMaterializedViewStmt stmt)
         // check db quota
         db.checkQuota();
 
-        db.writeLock();
+        if (!db.writeLockAndCheckExist()) {
+            throw new DdlException("create materialized failed. database:" + db.getFullName() + " not exist");
+        }
         try {
             Table table = db.getTable(tableName);
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:134]: should check null?

File: fe/fe-core/src/main/java/com/starrocks/alter/Alter.java [hunk 3]
```diff
@@ -341,24 +346,29 @@ public void replayChangeMaterializedViewRefreshScheme(ChangeMaterializedViewRefr
...
-                        "interval step to {}, timeunit to {}, id: {}",
-                oldMaterializedView.getName(), refreshType.name(), asyncRefreshContext.getStartTime(),
-                asyncRefreshContext.getStep(),
-                asyncRefreshContext.getTimeUnit(), oldMaterializedView.getId());
+        db.writeLock();
+        try {
+            MaterializedView oldMaterializedView;
+            final MaterializedView.MvRefreshScheme newMvRefreshScheme = new MaterializedView.MvRefreshScheme();
+
+            oldMaterializedView = (MaterializedView) db.getTable(id);
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:354]: check null?

File: fe/fe-core/src/main/java/com/starrocks/alter/Alter.java [hunk 4]
```diff
@@ -341,24 +346,29 @@ public void replayChangeMaterializedViewRefreshScheme(ChangeMaterializedViewRefr
...
-        newMvRefreshScheme.setType(refreshType);
-        newMvRefreshScheme.setAsyncRefreshContext(asyncRefreshContext);
-        oldMaterializedView.setRefreshScheme(newMvRefreshScheme);
-        LOG.info(
-                "Replay materialized view [{}]'s refresh type to {}, start time to {}, " +
-                        "interval step to {}, timeunit to {}, id: {}",
-                oldMaterializedView.getName(), refreshType.name(), asyncRefreshContext.getStartTime(),
-                asyncRefreshContext.getStep(),
-                asyncRefreshContext.getTimeUnit(), oldMaterializedView.getId());
+        db.writeLock();
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:349]: use new api?

Comment (Astralidea) [reply fe/fe-core/src/main/java/com/starrocks/alter/Alter.java:352]: I think in replay not does need to check db exists

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 18 / 46 (39.13%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/alter/Alter.java|4|27|14.81%|[131, 136, 167, 232, 352, 355, 357, 358, 359, 360, 362, 363, 364, 365, 366, 367, 368, 369, 370, 373, 374, 375, 377]|
|:large_blue_circle:|com/starrocks/scheduler/PartitionBasedMaterializedViewRefreshProcessor.java|14|19|73.68%|[153, 159, 554, 560, 564]|

**Review by ABingHuang — APPROVED** [review]
LGTM

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by decster — APPROVED** [review]

**Comment by Astralidea** [comment]
@Mergifyio backport branch-2.4

**Pull request merged**